Add Prompt Management documentation (#38005)
DataDog/documentation: content/en/llm_observability/monitoring/prompt_management.md

further_reading:
  - link: "/llm_observability/monitoring/prompt_tracking"
    tag: "Documentation"
    text: "Prompt Tracking"
  - link: "/llm_observability/playground"
    tag: "Documentation"
    text: "Playground"
  - link: "/llm_observability/instrumentation/sdk"
    tag: "Documentation"
    text: "Agent Observability SDK"

---

{{< callout url="https://www.datadoghq.com/" btn_hidden="true">}}
Prompt Management is in Preview.
{{< /callout >}}

## Overview

Prompt Management provides a centralized registry for the prompts used by your LLM applications. Instead of hardcoding prompt templates in application code or configuration files, create, version, and update prompts through Agent Observability, then fetch them at runtime.

This decouples prompt changes from your application's deployment cycle and allows non-technical stakeholders (PMs, SMEs) to test ideas without a code deployment.

Prompt Management works alongside [Prompt Tracking][1]: prompts fetched from the registry are automatically tagged on the LLM spans that use them, and their version history is visible in the Prompts view.

## Prerequisites

- A [Datadog API key][2] to fetch the prompts. 

> - An [application key][3] for write operations (creating or updating prompts). 

Reviewer: app key (or SAT) is actually needed for read operations (so get_prompt will need it) generally across datadog. 
we are working with ffe to find a work around, since this is a bit worse user experience, but its true right now

Author: @PROFeNoM is this still the case? (appkey needed for read)

Reviewer 2: yes

In the Prompt Editor:

1. Add one or more messages and assign each a role: {{< ui >}}System{{< /ui >}}, {{< ui >}}User{{< /ui >}}, or {{< ui >}}Assistant{{< /ui >}}.
2. Use `{{variable_name}}` syntax in any message to add dynamic content.
3. (Optional) Click {{< ui >}}Run{{< /ui >}} to test the prompt with sample values.
4. Click {{< ui >}}Save Prompt{{< /ui >}} to open the save dialog.

We recommend structuring your prompt so that the user query and context get injected as variable like so:

{{< img src="llm_observability/monitoring/prompt-creation.png" alt="Creating a Prompt in Agent Observability." style="width:100%;" >}}

In the save dialog:

| Field | Description |
|-------|-------------|
| {{< ui >}}Prompt ID{{< /ui >}} | A unique identifier for the prompt (for example, `customer-support-greeting`). Use this ID to fetch the prompt at runtime with `LLMObs.get_prompt()`. |
| {{< ui >}}Description{{< /ui >}} | Optional notes about this version. |
| {{< ui >}}Deployment{{< /ui >}} | The environment to deploy this version to. |

Click {{< ui >}}Create Prompt{{< /ui >}} to save the prompt to the registry.

> #### Via API

Reviewer 2: When creating via the API (whether it is direct HTTP hits or using the SDK methods), **the user will need an app key with the `llm_observability_write` scope** - I don't believe this is the case by default, and it would otherwise result in `Prompt API error (403): {"errors":["Forbidden","Failed permission authorization checks"]}`

Endpoint
: `https://api.{{< region-param key="dd_site" code="true" >}}/api/unstable/llm-obs/v1/prompts`

Method
: `POST`

Headers (required)
- `DD-API-KEY=<YOUR_DATADOG_API_KEY>`
- `DD-APPLICATION-KEY=<YOUR_DATADOG_APPLICATION_KEY>`
- `Content-Type="application/json"`

{{< code-block lang="bash" >}}
curl -X POST "https://api.datadoghq.com/api/unstable/llm-obs/v1/prompts" \
-H "DD-API-KEY: <YOUR_DATADOG_API_KEY>" \
-H "DD-APPLICATION-KEY: <YOUR_DATADOG_APPLICATION_KEY>" \
-H "Content-Type: application/json" \
-d '{
  "prompt_id": "customer-support-greeting",
  "title": "Customer support greeting",
  "description": "Initial greeting template used by the support bot",
  "template": [
    {"role": "system", "content": "You are a support agent for {{company}}."},
    {"role": "user", "content": "{{question}}"}
  ]
}'
{{< /code-block >}}

Creating a prompt with a `prompt_id` that already exists in the registry returns a `409` response.

### Update, list, and delete prompts

#### Via UI

Open a prompt in the {{< ui >}}Prompts{{< /ui >}} page to:

- **Create a new version**: {{< ui >}}Edit{{< /ui >}} the messages in the Prompt Editor to create a new version of an existing prompt.

> - **Deploy a version to a different environment**: select a version and update its {{< ui >}}Deployment{{< /ui >}} target to new environments.

Reviewer: im not 100% sure if we have update deployed env via http api/python sdk crud right now. @PROFeNoM to confirm either way?

Reviewer 2: Is your question about whether a user can change the environment in which a prompt is being "deployed" from the API?

Well...

Technically, given that prompts are managed feature flags, and managed feature flags do not prevent the user from changing allocations nor enabling a feature flag in an environment, they should technically be able to use the existing [Feature Flags API](https://docs.datadoghq.com/api/latest/feature-flags/) to manage the corresponding feature flag.

So, right now, indirectly, this should be able to be done.

But i haven't tested it...

- **Delete a prompt**: select {{< ui >}}Delete{{< /ui >}} from the prompt's options menu. This removes the prompt and its version history from the registry.

> #### Via API

Reviewer 2: Similarly to the comment above, the write endpoints will all need the app key to have the LLM Observability Write scope (i'd even add "Read" to be 100% safe, although it should work without it)

Endpoint
: `https://api.{{< region-param key="dd_site" code="true" >}}/api/unstable/llm-obs/v1/prompts/{prompt_id}`

Method
: `GET`

Headers (required)
- `DD-API-KEY=<YOUR_DATADOG_API_KEY>`

{{< code-block lang="bash" >}}
curl -X GET "https://api.datadoghq.com/api/unstable/llm-obs/v1/prompts/customer-support-greeting" \
-H "DD-API-KEY: <YOUR_DATADOG_API_KEY>"
{{< /code-block >}}

The raw HTTP endpoint always returns the latest version of a prompt. `DD_ENV`-based resolution is only available through the SDK's `get_prompt` method.

#### Fetch a specific version

Endpoint
: `https://api.{{< region-param key="dd_site" code="true" >}}/api/unstable/llm-obs/v1/prompts/{prompt_id}/versions/{version}`

Method
: `GET`

Headers (required)
- `DD-API-KEY=<YOUR_DATADOG_API_KEY>`

{{< code-block lang="bash" >}}
curl -s -X GET "https://api.datadoghq.com/api/unstable/llm-obs/v1/prompts/customer-support-greeting/versions/2" \
-H "DD-API-KEY: <YOUR_DATADOG_API_KEY>"
{{< /code-block >}}

## Monitor prompt usage

To see prompt metadata on the LLM spans it generates, Agent Observability must be enabled (with `LLMObs.enable()` or the equivalent environment variables).   

> For LLM calls made through an auto-instrumented provider integration (OpenAI, Anthropic, and others), passing the output of `prompt.format()` directly into the call tags the resulting span automatically, without extra instrumentation:

Reviewer: @PROFeNoM is this true? haven't read your auto tagging pr yet

Reviewer 2: Yes, this **happy** path should work.

> ---

Reviewer 3: Is this intended to be a private preview? If we don't want it to show in search, you can add `private: true` to the frontmattter. If it's not intended to be private, it is missing from the left-nav. You can add it there via `config/_default/menus/main.en.yaml`.

Reviewer 2: > Is this intended to be a private preview?

@CFLJacquet so FYI, I hence removed the Preview banner added from a [commit](https://github.com/DataDog/documentation/pull/38005/changes/4f513db3d951910ae94a1ede5ad909e78d39ae34)

I never really touched it until today although it felt weird (because afaik we're GAing?), but removed it in [this commit](https://github.com/DataDog/documentation/pull/38005/commits/880af64731b28aa1649999e91906fbfafa3a7208).

sgty?

Author: let's keep it public + keep the preview banner + add to left-nav. we are still in preview until we get 5 customers using it at scale. Customers need to be manually added to the preview => they have to reach out to their CSM that relays the info to us

title: Prompt Management
description: Create, version, and retrieve managed prompts in Python applications with Prompt Management.

further_reading:
  - link: "/llm_observability/monitoring/prompt_tracking"
    tag: "Documentation"
    text: "Prompt Tracking"
  - link: "/llm_observability/playground"
    tag: "Documentation"
    text: "Playground"
  - link: "/llm_observability/instrumentation/sdk/?tab=python"
    tag: "Documentation"
    text: "Agent Observability SDK"

---

> {{< callout url="https://www.datadoghq.com/" btn_hidden="true">}}

Reviewer 3: Does a user need to do anything special to get access to the preview?

Reviewer 2: Removed the banner in https://github.com/DataDog/documentation/pull/38005/commits/880af64731b28aa1649999e91906fbfafa3a7208 - See https://github.com/DataDog/documentation/pull/38005#discussion_r3671624440

Runtime retrieval is supported in Python through the `ddtrace` SDK. Prompt retrieval and Prompt Tracking are separate: `LLMObs.get_prompt()` can retrieve a managed prompt without enabling Agent Observability, but Agent Observability must be enabled to create LLM spans and associate prompt metadata with them.

Prompt Management works alongside [Prompt Tracking][1]. When Agent Observability is enabled, managed prompts passed directly to supported, automatically instrumented LLM calls are associated with the resulting spans.

## Prerequisites

- Python 3.9 or later.
- Your [Datadog site][2] and a [Datadog API key][3]. The API key is required for prompt retrieval even if traces are sent through the Datadog Agent.
- A [Datadog application key][4] with the `llm_observability_read`, `feature_flag_config_read`, and `feature_flag_environment_config_read` permissions to resolve prompts by environment. If you select an existing application key in Datadog, ensure that it has these permissions.
- To manage prompts through the API or Python SDK, the application key also requires the `llm_observability_write` and `feature_flag_config_write` permissions.

> ### Install the SDK

Reviewer 3: It's a bit more typical in our docs for this to not be a subheading under prerequisities.
```suggestion
## Install the SDK
```

Install or upgrade the latest `ddtrace` package in the Python environment used by your application:

```shell
pip install --upgrade ddtrace
```

## Use a managed prompt in Python

### Integrate Prompt Management with a coding agent

Integrate a managed prompt with a coding agent of your choice by pasting in the following prompt:

> ```bash

Reviewer 3: ```suggestion
```text
```

## Create and manage prompts

Create prompts and publish new versions in the {{< ui >}}Prompts{{< /ui >}} UI, through the Python SDK, or through the API.

### Create a prompt

#### Promote a tracked prompt

To promote a prompt already tracked in Agent Observability to a managed prompt, navigate to the {{< ui >}}Prompts{{< /ui >}} page, open the prompt, and click {{< ui >}}Register{{< /ui >}}. You can then update the prompt in the UI and retrieve it at runtime.

#### In the UI from scratch

Navigate to the {{< ui >}}Prompts{{< /ui >}} page and click {{< ui >}}+ New Prompt{{< /ui >}}.

In the Prompt Editor:

1. Add one or more messages and assign each a role: {{< ui >}}System{{< /ui >}}, {{< ui >}}User{{< /ui >}}, or {{< ui >}}Assistant{{< /ui >}}.
2. Use `{{variable_name}}` syntax in any message to add dynamic content.
3. Optional: Click {{< ui >}}Run{{< /ui >}} to test the prompt with sample values.
4. Click {{< ui >}}Save Prompt{{< /ui >}} to open the save dialog.

Structure the prompt so the user query and context are injected as variables:

> {{< img src="llm_observability/monitoring/prompt-creation.png" alt="Creating a Prompt in Agent Observability." style="width:100%;" >}}

Reviewer 3: ```suggestion
{{< img src="llm_observability/monitoring/prompt-creation.png" alt="The Playground with a System Prompt message reading 'You are a support agent for {{company}}' and a User Prompt message containing {{question}}, with the Save Prompt button in the top right." style="width:100%;" >}}
```

In the save dialog:

| Field | Description |
|-------|-------------|
| {{< ui >}}Prompt ID{{< /ui >}} | A unique identifier for the prompt, such as `customer-support-greeting`. Use this ID to retrieve the prompt with `LLMObs.get_prompt()`. |
| {{< ui >}}Description{{< /ui >}} | Optional notes about this version. |
| {{< ui >}}Deployment{{< /ui >}} | The environment to which this version is deployed. |

Click {{< ui >}}Create Prompt{{< /ui >}} to save the prompt to the registry.

### Update, list, and delete prompts

#### In the UI

Open a prompt in the {{< ui >}}Prompts{{< /ui >}} page to:

- **Create a new version**: Click {{< ui >}}Edit{{< /ui >}} and update the messages in the Prompt Editor.
- **Deploy a version to another environment**: Select a version and update its {{< ui >}}Deployment{{< /ui >}} environments.
- **Delete a prompt**: Select {{< ui >}}Delete{{< /ui >}} from the prompt's options menu. This removes the prompt and its version history from the registry.

### Use the Python SDK

> Use `LLMObs.create_prompt()` to create a prompt and deploy its first version to one or more environments. The `env_ids` values are Feature Flags environment IDs, which you can obtain from the [List environments API](/api/latest/feature-flags/list-environments/):

Reviewer 3: Please make this a reference-style link like the rest of the page. Thanks!

Reviewer 2: Added in https://github.com/DataDog/documentation/pull/38005/commits/16dd6c2731c9f049bd67e12a932a243df7b1081f
